File: Lua/Stack.lua
-- Stack "Class"
-- December 10, 2017
-- Implementation of the Stack ADT in Lua.

STACK = {}

-- Wrapper function that creates and returns the "Stack" _stack and methods.
function STACK:Create()

	local _stack = {}
	_stack.list = {}

	-- Returns the first element in _stack.list or nil.
	function _stack:top()
		if self.list[1] then
			return self.list[1]
		else
			return nil
		end
	end

	-- Pushes a new element into _stack.list.
	function _stack:push( pushed )
		if pushed then
			table.insert(self.list, pushed)
		end
	end

	-- Pops the element stored in _stack.list or nil.
	function _stack:pop()
		if self.list[1] then
			return table.remove(self.list, #self.list)
		else
			return nil
		end
	end

	return _stack
end

File: Java/Stack.java
/*
Stack<T> Class
December 10, 2017
Implementation of the Stack ADT in Java.
*/
public final class Stack<T> {
    
    /*
    E<T> Class
    Private class containing (E)lements of the Stack<T>
    */
    class E<T> {
        public T data = null;
        public E next = null;
        
        public E(T data, E next) {
            this.data = data;
            this.next = next;
        }
    }
    
    private E<T> top;
    
    // Initialize Stack with null top.
    public Stack() {
        this.top = null;
    }
    
    // Initialize Stack with an array of type T.
    public Stack(T[] array) {
        for(T element : array)
            this.push(element);
    }
    
    // Returns the data stored in the top or null if empty.
    public T top() {
        if(top == null)
            return null;
        else
            return top.data;
    }
    
    // Pushes a new element into the Stack.
    public T push(T pushed) {
        if(top == null)
            top = new E(pushed, null);
        else
            top = new E(pushed, top);
        return pushed;
    }
    
    // Pops the element stored in the top or null.
    public T pop() {
        if(top == null)
            return null;
        T popped = top.data;
        top = top.next;
        return popped;
    }   
}
